### PR TITLE
ssh2john.py: handle multiple keys

### LINKED ISSUE #4993: ssh2john.py doesn't extract all keys from a keyfile
Hello folks,

I'm having an issue where I can't crack a `format=ssh` rsa key... in fact, it's one grabbed from the vuln machine from basic pentesting room on tryhackme.

I'm running kali on WSL2 on Windows 10.
```
┌──(kali㉿DESKTOP-3TJKPJG)-[~/tryhackme/01_basic_pentesting]
└─$ uname -a
Linux DESKTOP-3TJKPJG 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 GNU/Linux
```

I've built the latest from the repo:
```
┌──(kali㉿DESKTOP-3TJKPJG)-[~/tryhackme/01_basic_pentesting]
└─$ /home/kali/john_from_github/john/run/john --list=build-info
Version: 1.9.0-jumbo-1+bleeding-c470f5d8b 2021-12-28 11:55:38 +0100
Build: linux-gnu 64-bit x86_64 AVX2 AC OMP
SIMD: AVX2, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
CPU tests: AVX2
$JOHN is /home/kali/john_from_github/john/run/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
SINGLE_IDX_MAX: 32768
SINGLE_BUF_MAX: 4294967295
Effective limit: Max. KPC 32768
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 11.2.0
GNU libc version: 2.32 (loaded: 2.32)
Crypto library: OpenSSL
OpenSSL library version: 0101010cf
OpenSSL 1.1.1l  24 Aug 2021
GMP library version: 6.2.1
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
times(2) sysconf(_SC_CLK_TCK) is 100
Using times(2) for timers, resolution 10 ms
HR timer: clock_gettime(), latency 100 ns
Total physical host memory: 12717 MiB
Available physical host memory: 12357 MiB
Terminal locale string: en_US.UTF-8
Parsed terminal locale: UTF-8
```

I've verified that the passphrase is `beeswax`, which is located in rockyou, by using the key for auth with `ssh -i`.

However, i can't crack using `john` as follows:
```
┌──(kali㉿DESKTOP-3TJKPJG)-[~/tryhackme/01_basic_pentesting]
└─$ /home/kali/john_from_github/john/run/ssh2john.py ./linpeas/kay_id_rsa > ./john/kay_id_rsa_john_hash

┌──(kali㉿DESKTOP-3TJKPJG)-[~/tryhackme/01_basic_pentesting]
└─$ /home/kali/john_from_github/john/run/john --log-stderr ./john/kay_id_rsa_john_hash --wordlist=/usr/share/wordlists/rockyou.txt
Using default input encoding: UTF-8
0:00:00:00 Starting a new session
0:00:00:00 Loaded a total of 1 password hash
Loaded 1 password hash (SSH, SSH private key [RSA/DSA/EC/OPENSSH 32/64])
0:00:00:00 Cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) is 0 for all loaded hashes
Cost 1 (KDF/cipher [0=MD5/AES 1=MD5/3DES 2=Bcrypt/AES]) is 0 for all loaded hashes
0:00:00:00 Cost 2 (iteration count) is 1 for all loaded hashes
Cost 2 (iteration count) is 1 for all loaded hashes
Will run 4 OpenMP threads
0:00:00:00 Command line: /home/kali/john_from_github/john/run/john --log-stderr ./john/kay_id_rsa_john_hash --wordlist=/usr/share/wordlists/rockyou.txt
0:00:00:00 - UTF-8 input encoding enabled
0:00:00:00 - Passwords will be stored UTF-8 encoded in .pot file
0:00:00:00 Autotune found best speed at OMP scale of 1
0:00:00:00 - Hash type: SSH, SSH private key (min-len 0, max-len 10 [worst case UTF-8] to 32 [ASCII])
0:00:00:00 - Algorithm: RSA/DSA/EC/OPENSSH 32/64
0:00:00:00 - Will reject candidates longer than 32 bytes
0:00:00:00 - Candidate passwords will be buffered and tried in chunks of 32
0:00:00:00 Proceeding with wordlist mode
0:00:00:00 - Wordlist file: /usr/share/wordlists/rockyou.txt
0:00:00:00 - memory mapping wordlist (139921507 bytes)
0:00:00:00 - No word mangling rules
Press 'q' or Ctrl-C to abort, almost any other key for status
0:00:00:00 - No stacked rules
0g 0:00:00:06 DONE (2022-01-09 22:30) 0g/s 2324Kp/s 2324Kc/s 2324KC/sa6_123..*7¡Vamos!
0:00:00:06 Session completed
Session completed.
```

I came across [an old issue](https://github.com/openwall/john/issues/4069) and none of the suggestions (like arg order, or using single versus double arg declarations) solve this issue.  Things seem to work fine... the wordlist is loading, `john` recognizes the `format` of the target, etc.

I feel like this is a bug, however, if I should post to the userlist let me know.

Thanks

[kay_id_rsa.txt](https://github.com/openwall/john/files/7836712/kay_id_rsa.txt)
[kay_id_rsa_john_hash.txt](https://github.com/openwall/john/files/7836713/kay_id_rsa_john_hash.txt)

### PR DESCRIPTION
**Pull request by pradkrish**
Closes #4993

### CODE REVIEW

**Comment by pradkrish** [comment]
I am opening the PR already to get some early feedback and work on it incrementally.

In the example given by @mbrownnycnyc the first key fails due to missing linefeeds. I am sure there are many other reasons why processing a key may fail. In order to keep it generic, the way I am handling currently is by signalling to the user with an error message that keys could be broken.

**Comment by solardiz** [comment]
@exploide You want to help review/test this? For the first commit, adding `?w=1` to the GitHub URL is helpful to see the actual changes (excluding whitespace changes).

@pradkrish Please prefix the commit messages with `ssh2john.py: `.

Thanks!

File: run/ssh2john.py [hunk 1]
```diff
@@ -45,6 +45,27 @@
     'AES-256-CTR': {'cipher': AES_256, 'keysize': 32, 'blocksize': 16, 'mode': "AES.MODE_CTR"},
 }
 
+def get_all_tags_ktypes(lines):
+    tags = []
+    ktypes = []
+    for line in lines:
+        if line.find("BEGIN RSA PRIVATE") > 0:
```
Comment (exploide) [inline run/ssh2john.py]: I think `if "BEGIN RSA PRIVATE" in line:` is more idiomatic Python. But I don't know how important comments on style are.

**Review by exploide — CHANGES_REQUESTED** [review]
Looks fine so far.

But this remark from @magnumripper is currently not taken into account:

> Perhaps it'd be a good idea to prefix the output hash(es) with a virtual login field? For a single hash output, it could be just the input filename as in kay_id_rsa.txt and for 2nd hash on it could be suffixed by an index such as kay_id_rsa.txt_2 and so on.

This means all hash lines are prefixed with the identical file name. This does not help to distinguish them, so better implement the index suffix.

File: run/ssh2john.py (hunk 1 above)
Comment (pradkrish) [reply run/ssh2john.py]: Idiomatic python is the way to go, thanks :)

Comment (pradkrish) [reply run/ssh2john.py]: Regarding the index suffix, it is definitely I plan to work on next, just wanted to get an initial feedback.

**Comment by solardiz** [comment]
The "add index suffix" commit looks good to me.
@exploide Can you also take a look?
@pradkrish Can you please prefix the commit messages with `ssh2john.py: ` and force-push?
Thanks!

**Comment by solardiz** [comment]
@pradkrish I meant prefix _all_ 3 commit messages here with `ssh2john.py: `, not just the last commit's. (You'll probably use `git rebase -i` for that.)

**Comment by magnumripper** [comment]
> > Perhaps it'd be a good idea to prefix the output hash(es) with a virtual login field? For a single hash output, it could be just the input filename as in kay_id_rsa.txt and for 2nd hash on it could be suffixed by an index such as kay_id_rsa.txt_2 and so on.
> 
> This means all hash lines are prefixed with the identical file name. This does not help to distinguish them, so better implement the index suffix.

Multiple keyfiles should be supported as well (not sure it is? It should be), so `ssh2john keyfile1 keyfile2` could produce the following:
```
keyfile1:hash
keyfile2:hash
keyfile2-2:hash
```

**Comment by pradkrish** [comment]
It should work for multiple key files. Let's say `keyfile1` has only one key and `keyfile2` has two keys, the output for `ssh2john keyfile1 keyfile2` will be

```
keyfile1: hash
keyfile2_1: hash
keyfile2_2: hash
```

File: run/ssh2john.py [hunk 2]
```diff
@@ -60,146 +81,140 @@ def read_private_key(filename):
...
-        return
+            sys.stderr.write("%s uses unsupported cipher, please file a bug!\n" % f.name)
+            return
 
-    sys.stdout.write("%s\n" % hashline)
+        sys.stdout.write("%s\n" % hashline)
+        start = end + 1
+        num_processed_keys += 1
 
+    if not num_processed_keys == len(tags):
```
Comment (exploide) [inline run/ssh2john.py]: Minor nit: `if not ... == ...` can be simplified to `if ... != ...`

**Review by exploide — APPROVED** [review]
LGTM

**Pull request merged**